monitor/src/main/java/edu/asu/diging/monitor/web/admin/ModifyRecipientController.java

```diff
@@ -0,0 +1,85 @@
+import edu.asu.diging.monitor.core.service.INotificationManager;
+import edu.asu.diging.monitor.web.admin.forms.AppForm;
+import edu.asu.diging.monitor.web.admin.forms.RecipientForm;
+
+@Controller
+public class ModifyRecipientController {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Autowired
+    private INotificationManager manager;
+
+    @Autowired
+    private IAppManager appManager;
+
+    @RequestMapping(value = "/admin/recipients/{email}/modify", method = RequestMethod.GET)
+    public String show(@PathVariable("email") String email, Model model) {
+        INotificationRecipient recipientDetails = manager.getRecipient(email);
+        RecipientForm recipientForm = new RecipientForm();
+        recipientForm.setEmail(recipientDetails.getEmail());
+        recipientForm.setName(recipientDetails.getName());
+        recipientForm.setApps(appManager.getApps().stream().map(a -> {
+            AppForm app = new AppForm();
+            app.setId(a.getId());
+            app.setName(a.getName());
+            return app;
+        }).collect(Collectors.toList()));
+
+        List<String> appsList = recipientDetails.getApps().stream().map(a -> {
+            return a.getId();
+        }).collect(Collectors.toList());
+
+        model.addAttribute("recipientForm", recipientForm);
+        model.addAttribute("recipientApps", appsList);
+        return "admin/recipients/modify";
+    }
+
+    @RequestMapping(value = "/admin/recipients/modify", method = RequestMethod.POST)
+    public String modify(@ModelAttribute RecipientForm recipientForm, RedirectAttributes redirectAttrs) {
+        if (recipientForm.getEmail() == null || recipientForm.getEmail().trim().isEmpty()) {
+            redirectAttrs.addFlashAttribute("show_alert", true);
+            redirectAttrs.addFlashAttribute("alert_type", "danger");
+            redirectAttrs.addFlashAttribute("alert_msg",
+                    "Recipient could not be stored. Please provide an email address.");
+            return "redirect:/admin/recipients/modify";
+        }
+        if (recipientForm.getAppIds() == null) {
+            recipientForm.setAppIds(new ArrayList<>());
+        }
+        try {
+            manager.modifyRecipient(recipientForm.getEmail(), recipientForm.getName(), recipientForm.getAppIds());
+            redirectAttrs.addFlashAttribute("show_alert", true);
+            redirectAttrs.addFlashAttribute("alert_type", "success");
+            redirectAttrs.addFlashAttribute("alert_msg", "Recipient was successfully registered.");
+            return "redirect:/admin/recipient/list";
+        } catch (NoEmailRecipientException e) {
+            redirectAttrs.addFlashAttribute("show_alert", true);
+            redirectAttrs.addFlashAttribute("alert_type", "danger");
+            redirectAttrs.addFlashAttribute("alert_msg", "No such email recipient exists.");
+            return "redirect:/";
+        }
+    }
+}

```

Comment:
this class is pretty much a copy of the addrecipientcontroller, isn't it? either the controller should be reused, or duplicate code should be extracted (maybe a superclass makes sense).

Reply:
I wanted to keep the controllers separate because if I used the add method of AddRecipientController, I would have to check if the recipient email already exists in the database and call addRecipient() or modifyRecipient() accordingly. But if I were to merge these two methods by getting rid of the EmailAlreadyRegistered exception (as described in the above comment), I could achieve this using just one controller class. I did not create a superclass either becuase the exceptions to be caught are different in both cases. Could you please advise on how to proceed in this case?